File: tests/fixtures/c/events.c
// Minimal BTF fixture for btf2go E2E tests.
// Compiled with `clang -target bpf -g -O2 -c events.c -o events.elf`.
// We do not include any kernel headers — btf2go only needs the BTF
// type graph emitted by `-g`, not actual eBPF program semantics.

typedef unsigned char  u8;
typedef unsigned short u16;
typedef unsigned int   u32;
typedef unsigned long long u64;

struct inner_t {
    u32 lo;
    u32 hi;
};

union payload_t {
    u64 raw;
    struct inner_t pair;
};

struct events_t {
    u8  kind;             // offset 0
    // 3 bytes implicit pad
    u32 pid;              // offset 4
    u8  flag_a : 1;       // bitfield run at byte 8
    u8  flag_b : 1;
    u8  prio   : 6;
    // 7 bytes implicit pad to 8-byte align ts
    u64 ts;               // offset 16
    char comm[16];        // offset 24
    union payload_t pay;  // offset 40
};

// Anchor the struct so it survives into the emitted BTF.
struct events_t __events_anchor;

File: tests/fixtures/zig/events.zig
// Minimal BTF fixture for btf2go E2E tests, Zig flavour.
// Compiled with `zig build-obj -target bpfel-linux-none -ODebug events.zig`
// (or via `make -C tests/fixtures zig`).
//
// Mirrors the C fixture intentionally so users can compare layouts
// across toolchains. Zig is `extern struct` for repr(C) and
// `packed union` for unions; the BTF should look the same as the
// clang-emitted version.

const InnerT = extern struct {
    lo: u32,
    hi: u32,
};

const PayloadT = extern union {
    raw: u64,
    pair: InnerT,
};

const EventsT = extern struct {
    kind: u8,
    // 3 bytes implicit pad
    pid: u32,
    // We don't try to express C bitfields in Zig — Zig has its own
    // packed-struct semantics. A single u8 stand-in keeps the
    // layout matching the Rust fixture.
    flags_and_prio: u8,
    // 7 bytes implicit pad to align ts
    ts: u64,
    comm: [16]u8,
    pay: PayloadT,
};

// Anchor the type so BTF survives. `export` puts the symbol in a
// real section (not stripped) and `extern` ensures it doesn't get
// const-folded away.
export const events_anchor: EventsT = EventsT{
    .kind = 0,
    .pid = 0,
    .flags_and_prio = 0,
    .ts = 0,
    .comm = [_]u8{0} ** 16,
    .pay = PayloadT{ .raw = 0 },
};
